### DISCUSSION TITLE
Can't make GUI apps with Codespaces.

### DISCUSSION THREAD

**Discussion by YusufSuleman** [community/community#17420] [Codespaces] [answered]
I can't make GUI apps with Codespaces. I can only make web or terminal based apps. I know that Codespaces is running Ubuntu. You just need to add a lightweight desktop environment to the Codespaces Image and have it be accessible. You could do this like Windows 365 or have it as a button on the sidebar to access a GUI environment. Thanks, if you do it.

**Comment by Ashwin-droid** [comment]
you can always download the output and run it on your local machine

**Reply by kuashio** [reply]
I have the same problem, and I'm using Python with tkinter, so there's no executable output file, or is there?

**Reply by heipiao233** [reply]
So then why I use Codespaces? Codespaces loses it value if we do that.

**Comment by Liero** [comment]
I think you might be interested in [Azure DevBox:](ttps://techcommunity.microsoft.com/t5/azure-developer-community-blog/introducing-microsoft-dev-box/ba-p/3412063) which is currently in beta

**Comment by joshspicer** [comment, marked as answer]
Hey!  I think you'd be interested in our `desktop-lite` script/dev container feature, which spins up a desktop environment that you can interact with through VNC (or a VNC web client)

https://github.com/microsoft/vscode-dev-containers/blob/main/script-library/docs/desktop-lite.md

The [VS Code source code repo](https://github.com/microsoft/vscode/tree/main/.devcontainer#quick-start---github-codespaces) uses this for developing VS Code - I think it might be useful to take a peek at how they use it if you'd like to set up something similar!

**Reply by josealvarez97** [reply]
Do you think something like that could be used to develop for Unity3D? @joshspicer

**Reply by jfullstackdev** [reply]
this is the same answer marked in the other post, 

https://github.com/orgs/community/discussions/17137

**Reply by ghost** [reply]
Any way to get a webcam to work with desktop lite?

**Reply by joshspicer** [reply]
Graphics-intensive workflows, as well as passing through hardware from the host machine, are things that will likely be difficult to do with the desktop-lite Feature.  I don't have any examples for either to share 😞 

There was [some experimental work in the past](https://github.com/orgs/community/discussions/47619) ([and this one](https://github.com/orgs/community/discussions/60759)) around passing through things like USB devices, but that work hasn't been prioritized for several reasons, eg: complexity and time.

**Comment by jfullstackdev** [comment]
just this time around, I really wanted to have a full Codespace only setup for my Laravel projects, even the DB is inside the container, and the UI is not a problem since it's web development not the traditional local UI like Java Swing components

well yes, I got it working and I'm so happy now because this means you can even code purely on tablet or smartphone, but there are still limitations you will be hitting when using this and maybe fixed in the future, 

like when you need to modify server settings, instead of directly editing the file because the changes will be gone the next time you open the Codespace, you need to do it using the RUN command

same thing happens when you already have, say, a temporary data on your database and you are testing something, then, say, you have to have a break then suddenly Codespace need to restart because of inactivity, all of those data will be gone, there are solutions but outside of Codespace 

https://stackoverflow.com/questions/69195913/cant-keep-postgres-data-persistent-using-github-codespaces-with-docker-compose

it can be part of the seeder but what if you are testing temporary data too for a new feature, that's not possible in the seeder
